### ISSUE TITLE
Documentation

### ISSUE THREAD

**Issue by rffontenelle** [readthedocs/readthedocs.org#9380]
## Details

* Read the Docs project URL: https://readthedocs.org/projects/{your_project_slug}/
* Build URL (if applicable): https://readthedocs.org/projects/{your_project_slug}/builds/{build_id}/
* Read the Docs username (if applicable): https://readthedocs.org/profiles/{your_rtd_username}/

## Expected Result

*A description of what you wanted to happen*

## Actual Result

*A description of what actually happened*

**Comment by rffontenelle** [comment]
I would like to highlight the frc-docs-translations workflow brought up in https://github.com/readthedocs/readthedocs.org/issues/4463#issuecomment-950789132 and next comments, a job nicely done by @Daltz333 and others, which could be used as inspiration for enabling a workflow for fetching translation from as well updating in Transifex.

**Comment by rffontenelle** [comment]
@mendax1234 Want want explanation or PR we start working on?

**Comment by mendax1234** [comment]
l just want to make sure that @Daltz333 uses a github workflow to pull and push the sources automatically?

**Comment by Daltz333** [comment]
I do not touch the sources. Reviewers approve and then they get pulled in twice a day.

**Comment by Daltz333** [comment]
```
The first thing l want to know is whether the pull cycle is fixed?
```
Yes. [Every 12 hours](https://github.com/wpilibsuite/frc-docs-translations/blob/09d2301cf2f5eccc94e48f0ee76f19756d3ca24c/.github/workflows/update-translations.yml#L9).
```
The second thing is what does the Reviewers here mean?
```
The language reviewers on Transifex. If I remember correctly, the permissions goes Translator -> Reviewer -> Language Coordinator -> Project Maintainer
```
How do you push sources to transifex?
```
We [push](https://github.com/wpilibsuite/frc-docs-translations/blob/09d2301cf2f5eccc94e48f0ee76f19756d3ca24c/scripts/update.sh#L26) using the Transifex CLI client alongside. This will likely change in the near future, as the new client stability is pretty awful...
```
how do you deal with the condition that there will be two files on transifex
```
The translations are copied (by enabling translation memory in Transifex) and the [old file is locked for 7 days](https://github.com/wpilibsuite/frc-docs-translations/blob/main/scripts/lock-translations.py). After 7 days, it is deleted from the project. 7 days gives enough time in the situation where a file is moved AND modified, allowing translators to retranslate the file.
```
the translation of the part you have modifed will disappear
```
Correct, this is by design. Context is extremely important when translating. Even changing a single word in a sentence can dramatically alter the meaning, or context. So when a sentence is updated in source, the translation gets wiped for that specific line.

**Comment by mendax1234** [comment]
> Yes. [Every 12 hours](https://github.com/wpilibsuite/frc-docs-translations/blob/09d2301cf2f5eccc94e48f0ee76f19756d3ca24c/.github/workflows/update-translations.yml#L9).

Thanks , so you only use a github workflow to pull the translations permitted by the Reviewer (not the Project Maintainer) from transifex? Besides , how to set people to these different titile (like Reviewer, project Maintainer)?

> We [push](https://github.com/wpilibsuite/frc-docs-translations/blob/09d2301cf2f5eccc94e48f0ee76f19756d3ca24c/scripts/update.sh#L26) using the Transifex CLI client alongside. This will likely change in the near future, as the new client stability is pretty awful...

l notice that this `push` method is also automatic, so when will the `update.sh` run and how does it run? Do you also set in other places?

> The translations are copied (by enabling translation memory in Transifex) and the [old file is locked for 7 days](https://github.com/wpilibsuite/frc-docs-translations/blob/main/scripts/lock-translations.py). After 7 days, it is deleted from the project. 7 days gives enough time in the situation where a file is moved AND modified, allowing translators to retranslate the file.

l notice that you have written another .py file for this lock_translation , so does it have conflicts with transifex memory? Since both two have the similiar function l think. Besides , how do you enable transifex memory in your project? Is it payment function?

> If so , how do you deal with this kind of situation. That's when someone is translating a specific file (we call it file A) . And file A happens to be the file you are to modify. And you modify it and push it to the transifex. So, unfortunately, the translator who translates this part will have a vain work. Or , after you push this file, the previous file will be in the `locked` files and can still be seen?

**Comment by rffontenelle** [comment]
> In addition, if you only modify a specific file and don't change its location , it seems that when you push it to transifex again , the translation of the part you have modifed will disappear, how do you deal with this situation?

Old unused translations in Transifex are suggested to the translator with the percentage of similarity. So, when new source string are pushed to Transifex (even for a new or renamed file), the translator will see that string untranslated and there will be a suggestion.

**Comment by mendax1234** [comment]
> > In addition, if you only modify a specific file and don't change its location , it seems that when you push it to transifex again , the translation of the part you have modifed will disappear, how do you deal with this situation?
> 
> Old unused translations in Transifex are suggested to the translator with the percentage of similarity. So, when new source string are pushed to Transifex (even for a new or renamed file), the translator will see that string untranslated and there will be a suggestion.

Is that true. If so , that's understandable. So, l don't need to worry about the modified file's translation will disappear forever?

**Comment by rffontenelle** [comment]
> > > In addition, if you only modify a specific file and don't change its location , it seems that when you push it to transifex again , the translation of the part you have modifed will disappear, how do you deal with this situation?
> > 
> > 
> > Old unused translations in Transifex are suggested to the translator with the percentage of similarity. So, when new source string are pushed to Transifex (even for a new or renamed file), the translator will see that string untranslated and there will be a suggestion.
> 
> Is that true? If so , that's understandable. So, l don't need to worry about the modified file's translation will disappear forever?

Once a string is translated, it is added to the Translation Memory (TM). For more info on this TM suggestions and also screenshot, see https://docs.transifex.com/translation/using-translation-memory-and-concordance-search

**Comment by mendax1234** [comment]
> > > > In addition, if you only modify a specific file and don't change its location , it seems that when you push it to transifex again , the translation of the part you have modifed will disappear, how do you deal with this situation?
> > > 
> > > 
> > > Old unused translations in Transifex are suggested to the translator with the percentage of similarity. So, when new source string are pushed to Transifex (even for a new or renamed file), the translator will see that string untranslated and there will be a suggestion.
> > 
> > 
> > Is that true? If so , that's understandable. So, l don't need to worry about the modified file's translation will disappear forever?
> 
> Once a string is translated, it is added to the Translation Memory (TM). For more info on this TM suggestions and also screenshot, see https://docs.transifex.com/translation/using-translation-memory-and-concordance-search

Got it , thanks! It seems that l don't have to configure the translation memory. Then what about hte lockdown files? Can you explain how that is done?

**Comment by mendax1234** [comment]
> ```
> The first thing l want to know is whether the pull cycle is fixed?
> ```
> 
> Yes. [Every 12 hours](https://github.com/wpilibsuite/frc-docs-translations/blob/09d2301cf2f5eccc94e48f0ee76f19756d3ca24c/.github/workflows/update-translations.yml#L9).
> 
> ```
> The second thing is what does the Reviewers here mean?
> ```
> 
> The language reviewers on Transifex. If I remember correctly, the permissions goes Translator -> Reviewer -> Language Coordinator -> Project Maintainer
> 
> ```
> How do you push sources to transifex?
> ```
> 
> We [push](https://github.com/wpilibsuite/frc-docs-translations/blob/09d2301cf2f5eccc94e48f0ee76f19756d3ca24c/scripts/update.sh#L26) using the Transifex CLI client alongside. This will likely change in the near future, as the new client stability is pretty awful...
> 
> ```
> how do you deal with the condition that there will be two files on transifex
> ```
> 
> The translations are copied (by enabling translation memory in Transifex) and the [old file is locked for 7 days](https://github.com/wpilibsuite/frc-docs-translations/blob/main/scripts/lock-translations.py). After 7 days, it is deleted from the project. 7 days gives enough time in the situation where a file is moved AND modified, allowing translators to retranslate the file.
> 
> ```
> the translation of the part you have modifed will disappear
> ```
> 
> Correct, this is by design. Context is extremely important when translating. Even changing a single word in a sentence can dramatically alter the meaning, or context. So when a sentence is updated in source, the translation gets wiped for that specific line.

Hi @Daltz333 . l have tried to use your automatic push and pull method on my docs and l have some questions about it. l notice that in your `update-translations.yml` , it needs `secrets.TX_TOKEN`. But after l run this workflow successfully on my repo, it doesn't ask me to enter the `secrets.TX_TOKEN`, is that a normal phenomenon? Besides, can you further explain the expected result of your automatic push and pull method? In my repo, it seems that after l configure the settings and modify a file and push it to origin main branch. But nothing are pushed to transifex. 

Thanks

**Comment by mendax1234** [comment]
l try to push and pull through github action automatically, but it goes wrong and below is the wrong message .l have checked my .transifexrc file and it is correct since l can use tx push -s from the local development environment. So what’s wrong with that? And here is my repo: [mendax1234/6940Swerve-docs: FRC Team 6940’s Swerve Documentation Powered by Read the Docs (github.com)](https://github.com/mendax1234/6940Swerve-docs)
```
tx INFO: Updating /home/runner/.transifexrc file...
tx INFO: Using TX_TOKEN environment variable. Credentials won't be saved to .transifexrc.
tx ERROR: Authorization Required
tx ERROR: 
Error: Authentication failed. Please make sure your credentials are valid.
For more information, visit:
https://docs.transifex.com/client/client-configuration#-transifexrc.
```
l have checked my .transifexrc file and it is correct since l can use tx push -s from the local development environment. So what’s wrong with that? And here is my repo: [mendax1234/6940Swerve-docs: FRC Team 6940’s Swerve Documentation Powered by Read the Docs (github.com)](https://github.com/mendax1234/6940Swerve-docs)

**Comment by mendax1234** [comment]
I have successfully completed this automatic work in my project and l want to know one thing that how do you update your `.pot` file? Update it manually? Or it can also be done automatically? @Daltz333

**Comment by rffontenelle** [comment]
@mendax1234

> Got it , thanks! It seems that l don't have to configure the translation memory. Then what about hte lockdown files? Can you explain how that is done?

I'm not very familiar with the lockdown processe by Transifex, but the concept I know is: the maintainer wants to work on the translation files and this might take some time, so any translation contributed during this process might get lost.

When I say lost is: the translator already translated a string and then it appears as not translated (with a 100% suggestion by the Translation Memory). So, rework.

So locking the resource is a great way to avoid frustrations and rework for the translators, due to maintaince time.

> I try to push and pull through github action automatically, but it goes wrong and below is the wrong message .
> 
> ```
> tx INFO: Updating /home/runner/.transifexrc file...
> tx INFO: Using TX_TOKEN environment variable. Credentials won't be saved to .transifexrc.
> tx ERROR: Authorization Required
> tx ERROR: 
> Error: Authentication failed. Please make sure your credentials are valid.
> For more information, visit:
> https://docs.transifex.com/client/client-configuration#-transifexrc.
> ```
> 
> I have checked my .transifexrc file and it is correct since l can use tx push -s from the local development environment. So what’s wrong with that? And here is my repo: [mendax1234/6940Swerve-docs: FRC Team 6940’s Swerve Documentation Powered by Read the Docs (github.com)](https://github.com/mendax1234/6940Swerve-docs)

When using TX_TOKEN, tx ignores the .transifexrc. This is useful so you don't have to set this file in a CI/CD environment.

Have you generated a token in https://www.transifex.com/user/settings/api/ ? Make sure TX_TOKEN has the whole notation `1/something`. E.g.: 1/1123lihrfwkjh12k3ug545913401301823

**Comment by mendax1234** [comment]
Thanks! l have already done all the automatic work successfully! And thanks wpilib for providing me with such good resources.

> I'm not very familiar with the lockdown process by Transifex, but the concept I know is: the maintainer wants to work on the translation files and this might take some time, so any translation contributed during this process might get lost.

When I say lost is: the translator already translated a string and then it appears as not translated (with a 100% suggestion by the Translation Memory). So, rework.

So locking the resource is a great way to avoid frustrations and rework for the translators, due to maintenance time.

Yes , you are right , lockdown is a good choice.

> When using TX_TOKEN, tx ignores the .transifexrc. This is useful so you don't have to set this file in a CI/CD environment.

Have you generated a token in https://www.transifex.com/user/settings/api/ ? Make sure TX_TOKEN has the whole notation 1/something. E.g.: 1/1123lihrfwkjh12k3ug545913401301823

Yes , l have generated  a TX_TOKEN , but l find that you also need to set it in your project's `secret` section. Just create a new environment variable in your `secrets` and name it as `TX_TOKEN` and then paste your TOKEN which you get from transifex.

In addition , l think l will write an instruction on how this automatic method works but l don't know whether @Daltz333 will mind my doing this since most of the code is written by WPILIB and l just change some parts to suit my project. l think this instruction will help you write your PR!

Best
Daniel

**Comment by rffontenelle** [comment]
What CI/CD method would RTD prefer for automatizing the translation workflow? CircleCI, GitHub Actions, ... ?

**Comment by mendax1234** [comment]
> What CI/CD method would RTD prefer for automatizing the translation workflow? CircleCI, GitHub Actions, ... ? @mendax1234

I use Github action and l think it is enough. But l find that in this RTD repo has both of them. :joy:
![image](https://user-images.githubusercontent.com/62934294/176981070-8b537e32-6ffa-42a5-a83a-abc6c5c32532.png)

**Comment by mendax1234** [comment]
> In addition , l think l will write an instruction on how this automatic method works but l don't know whether @Daltz333 will mind my doing this since most of the code is written by WPILIB and l just change some parts to suit my project. l think this instruction will help you write your PR!

@rffontenelle l write a brief instruction on how to use this automatic workflow. Hope that it can be regarded as  a reference. Here is the [link](https://6940swerve-docs.readthedocs.io/en/latest/docs/contributing/index.html#more-about-how-our-translation-flow-works)

**Comment by rffontenelle** [comment]
@Daltz333 Should the new Transifex client work with the current sphinx-intl somehow? I'm getting `Error: Cannot create resource, i18n type is unknown` when trying to tx-push, and `tx migrate` won't work either.

@mendax1234 I'll check it out this weekend, thanks for taking notes

**Comment by rffontenelle** [comment]
@Daltz333 If implemented the Python version, should `tx migrate` do the whole job for the new tx client?

**Comment by Daltz333** [comment]
I do not believe the new client properly respects language mapping, so there is additional work there.

**Comment by mendax1234** [comment]
> > @rffontenelle l write a brief instruction on how to use this automatic workflow. Hope that it can be regarded as a reference. Here is the [link](https://6940swerve-docs.readthedocs.io/en/latest/docs/contributing/how-it-works.html)
> 
> 
> 
> @mendax1234 I haven't thoroughly tested the instructions, but here some considerations to improve them:
> 
> 
> 
> ### LANG_TO_PULL and LANG_MAP
> 
> I recommend that you break "Change the LANG_TO_PULL and LANG_MAP to the language you want to be translated into.". First of all, I felt "to the language you want to be translated into" was direct to me as translator when should be to the project maintainer. So I'd rephrase breaking instructions like this:
> 
> 
> 
> > Change the LANG_TO_PULL to meet the list of languages of languages you want to pull from Transifex. E.g.:
> 
> >```
> 
> >LANG_TO_PULL=${1:-'fr_CA,es_MX,zh_CN,tr_TR,he_IL,pt'}
> 
> >```
> 
> > **NOTE:** Make sure to keep the leading `${1:-`. This is is a Bash parameter expansion that allows a specific value to be used instead of this whole list, which useful for calling update.sh with a specific language code.
> 
> >
> 
> > Set LANG_MAP with any language code mapping that is required. This way, The language will be pulled from Transifex not with its variant language code, but the mapped language code. E.g.:
> 
> > ```
> 
> > LANG_MAP='es_MX: es, fr_CA: fr, he_IL: he, tr_TR: tr'
> 
> > ```
> 
> 
> 
> ### update.sh's Step 3 format typo
> 
> 
> 
> It seems to have a format issue here: "... ./source/\`(where you \`conf.py locates ..."
> 
> 
> 
> ### .github/workflow/publish_main.yml
> 
> 
> 
> Sorry, I don't know about this file. If it is not required for translations then I feel this should removed and go straight to .github/workflow/update-translations.yml
> 
> 
> 
> ### .github/workflow/update-translations.yml
> 
> 
> 
> I missed information on [how to generate Transifex API token](https://docs.transifex.com/account/authentication) and [explanation of what are secrets (TX_TOKEN!) and how to add them to a project's settings](https://docs.transifex.com/account/authentication).

Thanks for reminding! I will fix it soon!

**Comment by mendax1234** [comment]
@Daltz333 
Hi, l notice that you have changed the frc-docs-translation project to the new API. However,  when l try to implement the same thing on my project. Everything goes well except that the transifex has locked all my resources.
<img width="458" alt="image" src="https://user-images.githubusercontent.com/62934294/194218888-a5b3833b-4359-4264-b2de-3451af56b77e.png">

l guess that the tx-push command isn't run. Can you help me with that? Thanks! And here is my project:https://github.com/mendax1234/6940Swerve-docs

**Comment by rffontenelle** [comment]
I'll leave this patch of .circleci/config.yml, because I haven't tested for months now and I'm afraid of loosing it:

```
  # requires TX_TOKEN secret set with a valid Transifex APIv3 personal token
  translation-update:
    docker:
      - image: 'cimg/python:3.10'
    steps:
      - run: git clone --depth 1 https://${DOCS_GITHUB_TOKEN}@github.com/readthedocs/readthedocs.org.git
      - run: pip3 install -r requirements/docs.txt -r requirements/pip.txt
      # use the sphinx-intl's first commit (yet unreleased) that supports Transifex APIv3
      - run: pip3 install https://github.com/sphinx-doc/sphinx-intl@6b3913fea5a8c9b819818b9b28b6d0d0b6f022b3
      - run: curl -o- https://raw.githubusercontent.com/transifex/cli/master/install.sh | bash
          working_directory: /usr/local/bin
      - run: cd docs
      - run: make gettext
      - run: rm -f .tx/config
      - run: sphinx-intl create-txconfig
      - run: sphinx-intl update-txconfig-resources --transifex-project-name=readthedocs-docs --locale-dir locale --pot-dir _build/locale
      - run: tx push --source --no-interactive --skip
      - run: tx pull --transltions --use-git-timestamps
      - run: git config credential.helper 'cache --timeout=120'
      - run: git config user.name 'github-actions[bot]'
      - run: git config user.email 'github-actions[bot]@users.noreply.github.com'
      - run: git commit --allow-empty -m "[skip ci] Update docs translations"
      # Push quietly to prevent showing the token in log
      - run: git push -q https://${DOCS_GITHUB_TOKEN}@github.com/giantswarm/docs.git master
```

**Comment by agjohnson** [comment]
Just a note here: we worked translation pull/push into our release process, which happens weekly. This is far off from daily or twice daily, but automating commits was deemed too noisy. We're still looking for a better workflow here. Transifex is a particular source of pain, and it's quite awful having the translation workflow so disconnected from our development workflow. Crowdin was suggested, but it's not clear if it fixes the issues we have with Transifex automation or not.

**Comment by rffontenelle** [comment]
@agjohnson Thanks for your input. I suggest checking [Weblate](https://www.weblate.org/), an open-source web-based translation platform that has a pull request option (among others options). This pull request option allows to have a "Update translation via Weblate" pull request with commits auto-updated whenever there is a translation update in the translation project set in Weblate, and the maintainers merge it whenever they want. For free-software projects, using Hosted Weblate is free of charge using "libre hosting" plan.

**Comment by silopolis** [comment]
Le mer. 25 janv. 2023 à 04:00, Anthony ***@***.***> a écrit :
>
> Thanks for the pointer!

Speaking of pointers, these should help you make yourself an idea of
Weblate possibilities:

* https://docs.weblate.org/en/latest/vcs.html#github-pull-requests
* https://docs.weblate.org/en/latest/admin/continuous.html#pushing-changes-from-weblate
* https://docs.weblate.org/en/latest/admin/continuous.html#lazy-commits
* https://docs.weblate.org/en/latest/admin/addons.html#addon-weblate-git-squash

**Comment by rffontenelle** [comment]
Was there any evolution on this topic (Fix integration with [readthedocs-docs project](https://app.transifex.com/readthedocs/readthedocs-docs/) in Transifex, migrating to Weblate or Crowdin, etc.) ?

**Comment by hwhsu1231** [comment]
Hello, Read the Docs Community,

I'm the author of the @localizethedocs organization, which focuses on manufacturing l10n projects for the documentation of open-source projects. For example:

- [localizethedocs/cmake-docs-l10n](https://github.com/localizethedocs/cmake-docs-l10n)
- [localizethedocs/conan-docs-l10n](https://github.com/localizethedocs/conan-docs-l10n)
- [localizethedocs/ros2-docs-l10n](https://github.com/localizethedocs/ros2-docs-l10n)
- [and more](https://projects.localizethedocs.org)

I've already developed an infrastructure for l10n projects suitable for almost all Sphinx-based documentation, using [Crowdin](https://crowdin.com/) as the TMS platform. Currently, the `rtd-user-docs-l10n` and `rtd-dev-docs-l10n` projects are in the incubation phase:

- https://github.com/localizethedocs/localizethedocs.github.io/issues/8
- https://github.com/localizethedocs/localizethedocs.github.io/issues/9

Based on the previous discussion in this issue, it seems that some contributors suggested replacing **Transifex** with **Weblate** or **Crowdin**, due to some frustrations encountered while using Transifex. For example, the following content is replied from @agjohnson 

> Just a note here: we worked translation pull/push into our release process, which happens weekly. This is far off from daily or twice daily, but automating commits was deemed too noisy. We're still looking for a better workflow here. Transifex is a particular source of pain, and it's quite awful having the translation workflow so disconnected from our development workflow. Crowdin was suggested, but it's not clear if it fixes the issues we have with Transifex automation or not.

Since both the `user` and `dev` documentation share the same [configuration](https://github.com/readthedocs/readthedocs.org/blob/9172d56d102f3fb64a54f5dd95516e66dfcce0cf/docs/conf.py#L40-L55), and from the content perspective, we only need to focus on maintaining translations of the latest version, I should be able to complete the incubation and publish both of these two projects **within a month (or within a half month)**.

What do you think? Does this infrastructure match the workflow you're looking for?
